### PR TITLE
GFC-453: add display width attribute to text components

### PR DESCRIPTION
**Pull request by co-category**
@lutnos , @josef-vlach 

Please can you feedback whether there is a better way to do this. I have followed the existing approach. However, since the Text type definition had to be changed, it broke the tests in about 50 places. I resolved these but it is concerning that such a change would have such far reaching consequences.

### CODE REVIEW

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/ComponentType.scala [hunk 1]
```diff
@@ -50,6 +51,14 @@ case object Address {
   val fields = (id: FormComponentId) => mandatoryFields(id) ++ optionalFields(id)
 }
 
+object DisplayWidthAttribute extends Enumeration {
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/sharedmodel/formtemplate/ComponentType.scala]: `DisplayWidthAttribute` sound unnecessary long, what about dropping word `Attribute` to stay just with: `DisplayWidth` ?

File: test/uk/gov/hmrc/gform/sharedmodel/JsonParseTestFormat.scala [hunk 2]
```diff
@@ -101,4 +98,28 @@ class JsonParseTestFormat extends Spec {
...
+
+    val displayWidthOptions = Table(
+      // format: off
+      ("displayWidth", "expected"),
+      (DisplayWidthAttribute.XS.toString.toLowerCase,  Text(ShortText, Value, DisplayWidthAttribute.XS)),
+      (DisplayWidthAttribute.S.toString.toLowerCase,   Text(ShortText, Value, DisplayWidthAttribute.S)),
+      (DisplayWidthAttribute.M.toString.toLowerCase,   Text(ShortText, Value, DisplayWidthAttribute.M)),
+      (DisplayWidthAttribute.L.toString.toLowerCase,   Text(ShortText, Value, DisplayWidthAttribute.L)),
+      (DisplayWidthAttribute.XL.toString.toLowerCase,  Text(ShortText, Value, DisplayWidthAttribute.XL)),
+      (DisplayWidthAttribute.XXL.toString.toLowerCase, Text(ShortText, Value, DisplayWidthAttribute.XXL))
```
Comment (josef-vlach) [inline test/uk/gov/hmrc/gform/sharedmodel/JsonParseTestFormat.scala]: Maybe to add test for wrong display width to see default case (`L`)

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/ComponentType.scala (hunk 1 above)
Comment (co-category) [reply app/uk/gov/hmrc/gform/sharedmodel/formtemplate/ComponentType.scala]: Yes, DisplayWidth looks better

File: test/uk/gov/hmrc/gform/sharedmodel/JsonParseTestFormat.scala (hunk 2 above)
Comment (co-category) [reply test/uk/gov/hmrc/gform/sharedmodel/JsonParseTestFormat.scala]: Yes, will do

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/ComponentType.scala [hunk 3]
```diff
@@ -20,12 +20,13 @@ import cats.data.NonEmptyList
...
 import play.api.data.validation.ValidationError
 import play.api.libs.json._
+import uk.gov.hmrc.gform.sharedmodel.formtemplate.DisplayWidthAttribute.DisplayWidthAttribute
 
 import scala.collection.immutable._
 
 sealed trait ComponentType
 
-case class Text(constraint: TextConstraint, value: Expr) extends ComponentType
+case class Text(constraint: TextConstraint, value: Expr, displayWidth: DisplayWidthAttribute) extends ComponentType
```
Comment (lutnos) [inline app/uk/gov/hmrc/gform/sharedmodel/formtemplate/ComponentType.scala]: Would it be appropriate to add default value for displayWidth, do you think?  That would reduce the extra verbosity in many of the unit tests and reduce the overall size of the change

Comment (co-category) [reply app/uk/gov/hmrc/gform/sharedmodel/formtemplate/ComponentType.scala]: yes that's a good idea

File: app/uk/gov/hmrc/gform/formtemplate/FormComponentMaker.scala [hunk 4]
```diff
@@ -141,32 +143,37 @@ class FormCompomentMaker(json: JsValue) {
...
+      result <- (maybeFormatExpr, maybeValueExpr, multiline, displayWidth) match {
                  // format: off
-        case (Some(TextFormat(UkSortCodeFormat)), HasTextExpression(expr), IsNotMultiline()) => UkSortCode(expr).asRight
-        case (Some(TextFormat(f)),                HasTextExpression(expr), IsNotMultiline()) => Text(f, expr).asRight
-        case (None,                               HasTextExpression(expr), IsNotMultiline()) => Text(ShortText, expr).asRight
-        case (Some(TextFormat(f)),                HasTextExpression(expr), IsMultiline()   ) => TextArea(f, expr).asRight
-        case (None,                               HasTextExpression(expr), IsMultiline()   ) => TextArea(BasicText, expr).asRight
-        case (maybeInvalidFormat,                 maybeInvalidValue,       IsMultiline()   ) =>
+        case (Some(TextFormat(UkSortCodeFormat)), HasTextExpression(expr), IsNotMultiline(), _)                             => UkSortCode(expr).asRight
+        case (_, _, _, HasDisplayWidth(dw)) if (dw.toString == "INVALID")                                                   => UnexpectedState("Unsupported type of display width").asLeft
```
Comment (lutnos) [inline app/uk/gov/hmrc/gform/formtemplate/FormComponentMaker.scala]: Wouldn't this be better as: if (dw == DisplayWidth.INVALID)

**Pull request merged**